### PR TITLE
Feature: Liquid Staking

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: native-sdk/src/resource/resource_manager.rs [hunk 1]
```diff
@@ -1,13 +1,144 @@
...
+    ) -> Result<T, E>
+    where
+        Y: EngineApi<E> + Invokable<ResourceManagerGetNonFungibleInvocation, E>,
+    {
+        let output = api.invoke(ResourceManagerGetNonFungibleInvocation {
+            id,
+            receiver: self.0,
+        })?;
+
+        let data = scrypto_decode(&output[0]).unwrap();
```
Comment (iamyulong) [inline native-sdk/src/resource/resource_manager.rs:117]: `[0]` for mutable data only?

File: radix-engine-interface/src/api/types/actor.rs [hunk 2]
```diff
@@ -412,7 +413,18 @@ impl EpochManagerPackage {
...
+                        ))
+                    }
+
+                    ValidatorFn::ClaimXrd => {
+                        let args: ValidatorClaimXrdMethodArgs =
+                            scrypto_decode(args).map_err(ResolveError::DecodeError)?;
+                        NativeInvocation::Validator(ValidatorInvocation::ClaimXrd(
+                            ValidatorClaimXrdInvocation {
+                                receiver,
+                                bucket: args.bucket,
```
Comment (iamyulong) [inline radix-engine-interface/src/api/types/actor.rs:427]: Can we rename it to be more specific, like `unstake_nft` ?

File: radix-engine/src/model/epoch_manager/validator.rs [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+            let mut xrd_vault = Vault(validator.stake_xrd_vault_id);
+
+            let total_lp_supply = lp_token_resman.total_supply(api)?;
+            let active_stake_amount = xrd_vault.sys_amount(api)?;
+            let xrd_bucket = self.1;
+            let stake_amount = xrd_bucket.sys_amount(api)?;
+            let lp_mint_amount = if active_stake_amount.is_zero() {
+                stake_amount
+            } else {
+                stake_amount * total_lp_supply / active_stake_amount
```
Comment (iamyulong) [inline radix-engine/src/model/epoch_manager/validator.rs:215]: Blue screen here when `active_stake_amount == 0`?

File: radix-engine/src/model/epoch_manager/validator.rs [hunk 4]
```diff
@@ -0,0 +1,658 @@
...
+            let mut stake_vault = Vault(validator.stake_xrd_vault_id);
+            let mut unstake_vault = Vault(validator.pending_xrd_withdraw_vault_id);
+            let mut nft_resman = ResourceManager(validator.unstake_nft);
+            let mut lp_token_resman = ResourceManager(validator.liquidity_token);
+
+            let active_stake_amount = stake_vault.sys_amount(api)?;
+            let total_lp_supply = lp_token_resman.total_supply(api)?;
+            let lp_tokens = self.1;
+            let lp_token_amount = lp_tokens.sys_amount(api)?;
+            let xrd_amount = lp_token_amount * active_stake_amount / total_lp_supply;
```
Comment (iamyulong) [inline radix-engine/src/model/epoch_manager/validator.rs:307]: Panic here when `total_lp_supply == 0`?

File: radix-engine-interface/src/api/types/actor.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine-interface/src/api/types/actor.rs:427]: Yup

**Review by iamyulong — APPROVED** [review]

**Pull request merged**